### PR TITLE
Update ufs-bundle to build with head of develop of ufs-weather-model for ATM configuration

### LINKED ISSUE #14: Add the capability to run from UFS coldstart initial conditions
## Description

PR https://github.com/JCSDA/ufs-bundle/pull/9 adds the capability to do UFS ATM only warmstart runs. To do coldstart runs, the I/O layer in fv3-jedi needs to be modified and the test yamls updated.

### PR DESCRIPTION
**Pull request by mark-a-potts**
## Description

This PR, in combination with PRs for oops and fv3-jedi, allows the ufs-bundle to be built using the head of develop for the ufs-weather-model.  Additionally, new regression tests have been added to fv3-jedi that will update the run directories needed for the UFS model and to run a sample Ensemble Forecast of the UFS. 

## Issue(s) addressed

Resolves #14 

## Dependencies

List the other PRs that this PR is dependent on:
- [x] https://github.com/JCSDA-internal/fv3-jedi/pull/1099

## Impact

None. 

## Checklist

- [X] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [X] I have run the unit tests before creating the PR

### CODE REVIEW

**Comment by mark-a-potts** [comment]
This PR has been tested and works with UFS_APP=ATM, ATMAERO, and S2S configurations.  There are no tests for ATMAERO or S2S, but both configurations build. The NG-GODAS option does not build because the soca branch required is out of date with the rest of the bundle and requires fairly significant work.

**Comment by climbfuji** [comment]
> This PR has been tested and works with UFS_APP=ATM, ATMAERO, and S2S configurations. There are no tests for ATMAERO or S2S, but both configurations build. The NG-GODAS option does not build because the soca branch required is out of date with the rest of the bundle and requires fairly significant work.

I have an updated SOCA branch that I used for a similar effort ... I can make that work with this PR.

**Comment by mark-a-potts** [comment]
That would be great.  I started on a merge and it looked like it was going to be a lot of work. I added a patch to handle an issue with checksum checking of restart files down in atmos_cubed_sphere. I know that there are differences in the MOM6 repo, so maybe we could add a similar patch for those and get everything working with the head of develop of the WM.

**Comment by climbfuji** [comment]
@mark-a-potts Let's fix the CI failures first. This is because of a merge last week on the JEDI side that changed the model interfaces, but nobody thought about the fv3-jedi etc branches for the UFS. Let me do this in an hour or so, ok?

**Comment by mark-a-potts** [comment]
I have updated this to now handle building ATM, ATMAERO, and S2S using the head of develop of the ufs-wm. NG-GODAS is still very broken, but the build with S2S will actually run the ATM regression tests without issue.

**Comment by climbfuji** [comment]
@mark-a-potts I resolved the merge conflict in https://github.com/JCSDA/ufs-bundle/pull/43/commits/dd72818159051e4e0958ef0d9a6ce9df78b1ff43, please take a look to see if everything looks ok

File: README.md
Comment (climbfuji) [inline README.md]: We need to update this section (and the list of modules) to something more recent

**Comment by climbfuji** [comment]
There is now a CI failure in the cmake step, related to fv3-jedi-data: https://github.com/JCSDA/ufs-bundle/actions/runs/7133595917/job/19426615092?pr=43

**Comment by mark-a-potts** [comment]
Just removed the extra fv3-jedi-data section. I think it should work now.

**Comment by climbfuji** [comment]
Sigh, there are so many updates to the JEDI code that it seems impossible to get a clean run! I'll hop on the instance and clear the build area, then restart the test.

**Comment by mark-a-potts** [comment]
The latest update to ioda seems to have broken the buid now.

**Comment by climbfuji** [comment]
> The latest update to ioda seems to have broken the buid now.

Can you try to pull oops develop into your branch please?

**Comment by mark-a-potts** [comment]
Yeah, I think that was the issue. Just re-built successfully on my machine, so pushing now.

File: CMakeLists.txt
Comment (climbfuji) [inline CMakeLists.txt]: Note that this is not how we want to handle those switches in the future. Instead, the logic in the bundles should be like here: https://github.com/JCSDA-internal/jedi-bundle/blob/551edcbc73700544994c2b10fa1270fa25c1900a/CMakeLists.txt#L51C1-L51C91

File: CMakeLists.txt
Comment (climbfuji) [inline CMakeLists.txt]: Is this still needed, given that we updated the soca branch?

**Comment by climbfuji** [comment]
@mark-a-potts I am also running into problems when having to rerun make. Whether UPDATE_DISCONNECTED is ON or OFF, I get something like this:
```
[  0%] Performing update step for 'ufs-weather-model'
cd /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/ufs-weather-model && /opt/homebrew/Cellar/cmake/3.27.0/bin/cmake -Dcan_fetch=YES -P /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake
No stash entries found.
No stash entries found.
No stash entries found.
CMake Error at /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake:277 (message):


  Failed to unstash changes in:
  '/Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/ufs-weather-model'.


  You will have to resolve the conflicts manually


make[2]: *** [ufs-weather-model/src/ufs-weather-model-stamp/ufs-weather-model-update] Error 1
make[1]: *** [CMakeFiles/ufs-weather-model.dir/all] Error 2
make: *** [all] Error 2
```
This may have to do with the patch that gets applied to the ufs-weather-model develop branch?

**Comment by climbfuji** [comment]
> @mark-a-potts I am also running into problems when having to rerun make. Whether UPDATE_DISCONNECTED is ON or OFF, I get something like this:
> 
> ```
> [  0%] Performing update step for 'ufs-weather-model'
> cd /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/ufs-weather-model && /opt/homebrew/Cellar/cmake/3.27.0/bin/cmake -Dcan_fetch=YES -P /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake
> No stash entries found.
> No stash entries found.
> No stash entries found.
> CMake Error at /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake:277 (message):
> 
> 
>   Failed to unstash changes in:
>   '/Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/ufs-weather-model'.
> 
> 
>   You will have to resolve the conflicts manually
> 
> 
> make[2]: *** [ufs-weather-model/src/ufs-weather-model-stamp/ufs-weather-model-update] Error 1
> make[1]: *** [CMakeFiles/ufs-weather-model.dir/all] Error 2
> make: *** [all] Error 2
> ```
> 
> This may have to do with the patch that gets applied to the ufs-weather-model develop branch?

Update. I did a clean checkout and ran into the same problem:
```
[  0%] Performing disconnected update step for 'ufs-weather-model'
cd /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/ufs-weather-model && /opt/homebrew/Cellar/cmake/3.27.0/bin/cmake -Dcan_fetch=NO -P /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake
No stash entries found.
No stash entries found.
No stash entries found.
CMake Error at /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake:277 (message):


  Failed to unstash changes in:
  '/Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/ufs-weather-model'.


  You will have to resolve the conflicts manually


make[2]: *** [ufs-weather-model/src/ufs-weather-model-stamp/ufs-weather-model-update_disconnected] Error 1
make[1]: *** [CMakeFiles/ufs-weather-model.dir/all] Error 2
make: *** [all] Error 2
```
Since the patch only applies to the top-level ufs-weather-model and not to any of the submodules underneath, I would say we use a branch for the moment. The ufs-weather-model developers have indicated that they are open to these changes, therefore I expect that we wouldn't have to maintain that ufs-weather-model branch for very long.

File: CMakeLists.txt
Comment (mark-a-potts) [reply CMakeLists.txt]: If we switch to the branch I used for the PR into the WM, we can probably combine all APPs together, though there will still be a patch required to get ng-godas to build. Since the patches tend to be kind of fragile though, it might be best to only apply the patch for UFS_APP=NG-GODAS.

**Comment by mark-a-potts** [comment]
I cleaned up the CMakeLists.txt file a bit and fixed the *-data bundle syntax to match the other bundles. Note that the ufs-weather-model is now pointing to my PR branch (feature/fortran-flags) in github.com/NOAA-EPIC/ufs-weather-model. I got rid of the mom6 patch and added an ng-godas patch instead. It gets run at the top level of ufs-weather-model, but it is actually patching the MOM6 repo under MOM6-interface.  I have successfully built using UFS_APP=NG-GODAS and S2S, but still waiting on ATM and ATMAERO.

**Comment by climbfuji** [comment]
> I cleaned up the CMakeLists.txt file a bit and fixed the *-data bundle syntax to match the other bundles. Note that the ufs-weather-model is now pointing to my PR branch (feature/fortran-flags) in github.com/NOAA-EPIC/ufs-weather-model. I got rid of the mom6 patch and added an ng-godas patch instead. It gets run at the top level of ufs-weather-model, but it is actually patching the MOM6 repo under MOM6-interface. I have successfully built using UFS_APP=NG-GODAS and S2S, but still waiting on ATM and ATMAERO.

That sounds great, thanks very much @mark-a-potts

**Comment by climbfuji** [comment]
@mark-a-potts Looks like CI tests are passing now !

**Comment by mark-a-potts** [comment]
They didn't actually all run @climbfuji  But, it is close now. I will try submitting again.

**Comment by mark-a-potts** [comment]
This time the tests actually ran and seem to have passed. Yay!

File: CMakeLists.txt
Comment (climbfuji) [inline CMakeLists.txt]: I think we should keep the `UPDATE` here as well?

**Review by climbfuji — COMMENTED** [review]
I think this PR is ready for review, isn't it?

File: CMakeLists.txt
Comment (mark-a-potts) [reply CMakeLists.txt]: The soca branch needs some updates as well. I will put in a PR to feature/ufs_dom

**Comment by climbfuji** [comment]
@mark-a-potts I am unable to build any of the APPs with this PR. NO matter what I do, I get this from the `make` command after a fresh checkout and fresh cmake run:
```
[  0%] Performing disconnected update step for 'ufs-weather-model'
cd /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ufs-ensforcasts/ufs-weather-model && /opt/homebrew/Cellar/cmake/3.27.0/bin/cmake -Dcan_fetch=NO -P /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ufs-ensforcasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake
No stash entries found.
No stash entries found.
No stash entries found.
CMake Error at /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ufs-ensforcasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake:277 (message):


  Failed to unstash changes in:
  '/Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ufs-ensforcasts/ufs-weather-model'.


  You will have to resolve the conflicts manually


make[2]: *** [ufs-weather-model/src/ufs-weather-model-stamp/ufs-weather-model-update_disconnected] Error 1
make[1]: *** [CMakeFiles/ufs-weather-model.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by mark-a-potts** [comment]
> @mark-a-potts I am unable to build any of the APPs with this PR. NO matter what I do, I get this from the `make` command after a fresh checkout and fresh cmake run:
> 
> ```
> [  0%] Performing disconnected update step for 'ufs-weather-model'
> cd /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ufs-ensforcasts/ufs-weather-model && /opt/homebrew/Cellar/cmake/3.27.0/bin/cmake -Dcan_fetch=NO -P /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ufs-ensforcasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake
> No stash entries found.
> No stash entries found.
> No stash entries found.
> CMake Error at /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ufs-ensforcasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake:277 (message):
> 
> 
>   Failed to unstash changes in:
>   '/Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ufs-ensforcasts/ufs-weather-model'.
> 
> 
>   You will have to resolve the conflicts manually
> 
> 
> make[2]: *** [ufs-weather-model/src/ufs-weather-model-stamp/ufs-weather-model-update_disconnected] Error 1
> make[1]: *** [CMakeFiles/ufs-weather-model.dir/all] Error 2
> make: *** [all] Error 2
> ```

So, this might be because you need to remove the old ufs-weather-model directory? To accommodate the addition of the fortran flags for building, I switched to a fork of the WM under NOAA-EPIC, so if it still thinks it is looking for the branch under ufs-community, it will fail.

**Comment by mark-a-potts** [comment]
I just checked with @jkbk2004, and it sounds like my PR to the UFS-WM should go in later this week, but in the meantime, I extended the patch and will push up a new CMakeLists.txt file that pulls from ufs-community instead of NOAA-EPIC.

**Review by climbfuji — CHANGES_REQUESTED** [review]
I know why I get this failure even for a completely fresh checkout. It's because of this: https://github.com/NOAA-EMC/AQM/pull/40

So, long story short, the patch logic needs to be removed or otherwise we can't build/test on macOS any longer.

**Comment by mark-a-potts** [comment]
If it is exclusively a case-sensitivity issue, this might be an alternative for your testing--

https://brianboyko.medium.com/a-case-sensitive-src-folder-for-mac-programmers-176cc82a3830

We should be able to remove the patch to fix the CMakeLists.txt files when my PR into the WM gets merged, but the rest of that patch fixes issues in MOM6 that we need for coupling to SOCA. Given that the NG-GODAS coupling is broken right now anyway, I would probably lean towards removing the NG-GODAS option from ufs-bundle for the time being.  

We could also put in a PR to MOM6, but I feel kind of uncomfortable doing that if the SOCA connection doesn't really work right now.

**Comment by climbfuji** [comment]
I agree, let's hold off with the PR to mom6 until we know it is working. I was able to build it on my macOS using the case-sensitive volume trick you mentioned above. That's not ideal, but I don't see any better option either. Definitely we need to document this in the top-level `README.md` of this repository (i.e in this PR).

**Review by climbfuji — COMMENTED** [review]
This works for me if I am using a case-sensitive volume on my macOS. We need to get the dependent PRs approved before we can approve and merge this.

**Comment by mark-a-potts** [comment]
The oops PR is going to take a while, but I think I can comment out the ensemble test in fv3-jedi and get this working. I'll try that today.

File: CMakeLists.txt
Comment (climbfuji) [inline CMakeLists.txt]: Why does this need to be moved from above ioda to below ioda?

File: CMakeLists.txt
Comment (climbfuji) [inline CMakeLists.txt]: We should be able to revert the changes in line 60 and 61

File: README.md
Comment (climbfuji) [inline README.md]: Please also provide some background information and the link for creating the case-sensitive volume on macOS to compile w/ the patch logic.

**Review by climbfuji — COMMENTED** [review]
We are getting there, thanks for all your work!

File: CMakeLists.txt
Comment (mark-a-potts) [reply CMakeLists.txt]: Fixed this.

File: CMakeLists.txt
Comment (mark-a-potts) [reply CMakeLists.txt]: I think this is reverted now, too.

File: README.md
Comment (mark-a-potts) [reply README.md]: Done.

**Review by climbfuji — DISMISSED** [review]
Thanks very much for the `README.md` updates and your patience along the way!

**Comment by climbfuji** [comment]
Last thing is to get the fv3-jedi PR merged, correct? I think you avoided the oops dependency (it's still in the PR description) for now.

**Comment by mark-a-potts** [comment]
Yes. The fv3-jedi PR can be done without the oops PR, and I think it would be good to do that. I will remove the mentions of oops in the PR descriptions.

File: CMakeLists.txt
Comment (climbfuji) [inline CMakeLists.txt]: I still see a branch `feature/ufs-EnsForecasts` for fv3-jedi-data ?

File: CMakeLists.txt
Comment (climbfuji) [inline CMakeLists.txt]: The fv3-jedi branch `feature/ufs-EnsForecasts` was just merged into `develop`

File: CMakeLists.txt
Comment (climbfuji) [inline CMakeLists.txt]: Has this been merged already?

**Review by climbfuji — COMMENTED** [review]
Please update the PR from ufs-bundle develop, revert the fv3-jedi branch name; this PR also uses temporary branches for soca and fv3-jedi-data?

**Comment by climbfuji** [comment]
There is also a conflict for `README.md` because my last PR replaced `w3nco` with `w3emc`.

**Comment by mark-a-potts** [comment]
Working on this now. The s3 bucket is currently being synched to the latest rt results, so that is delaying the testing a bit. Will update shortly.

**Comment by mark-a-potts** [comment]
After the latest update, the ctests pass and builds now work for ATM, ATMAERO, NG-GODAS, and S2S configurations.

File: .github/workflows/run_ec2_pcluster.yaml
Comment (climbfuji) [inline .github/workflows/run_ec2_pcluster.yaml]: Why is this needed?

File: CMakeLists.txt
Comment (climbfuji) [inline CMakeLists.txt]:
```suggestion
ecbuild_bundle( PROJECT ufo-data GIT "https://github.com/jcsda-internal/ufo-data.git" BRANCH develop UPDATE )
```

File: CMakeLists.txt
Comment (climbfuji) [inline CMakeLists.txt]:
```suggestion
```

File: CMakeLists.txt
Comment (climbfuji) [inline CMakeLists.txt]:
```suggestion
```

File: README.md
Comment (climbfuji) [inline README.md]: This is reverting my recent change in `README.md` that replaced `w3nco` with `w3emc`

**Review by climbfuji — COMMENTED** [review]
A few minor comments to reduce the amount of unnecessary changes etc, other than that this looks great!

File: .github/workflows/run_ec2_pcluster.yaml
Comment (mark-a-potts) [reply .github/workflows/run_ec2_pcluster.yaml]: I had to put it there to make sure the build was started fresh. I'll remove it now.

File: README.md
Comment (mark-a-potts) [reply README.md]: Ah, I missed that. I mostly wanted to update the fms version to 2023.02.01 which is now required by the WM. I'll get that straightened out.

**Review by climbfuji — DISMISSED** [review]
Thanks for all your work on this!

**Comment by climbfuji** [comment]
@mark-a-potts CI tests are failing unfortunately: https://github.com/JCSDA/ufs-bundle/actions/runs/7745452671/job/21121475447?pr=43

**Review by climbfuji — APPROVED** [review]
Would have been nice to keep the build directories (much faster), but if it doesn't work ...

**Pull request merged**
